PSMT/File.pm

```diff
@@ -809,41 +750,62 @@ sub RegNewDoc {
     my ($self, $pathid, $name, $desc, $secure) = @_;
     my $docid = 0;
     my $dbh = PSMT->dbh;
-    $dbh->db_lock_tables('path WRITE', 'docreg WRITE');
+    $dbh->db_transaction_start();
     $self->ValidateNameInPath($pathid, $name);
     my $sth = $dbh->prepare('INSERT INTO docreg (pathid, filename, description, secure) VALUES (?, ?, ?, ?)');
-    if ($sth->execute($pathid, $name, $desc, $secure) == 0) {return $docid; }
+    if ($sth->execute($pathid, $name, $desc, $secure) == 0) {
+        $dbh->db_transaction_rollback();
+        return $docid;
+    }
     $docid = $dbh->db_last_key('docreg', 'docid');
-    $dbh->db_unlock_tables();
+    $dbh->db_transaction_commit();
 # No Need To Send Mail in 'RegNewDoc' : Always Followed by 'RegNewFile' !
 #    PSMT->email()->NewDocInPath($pathid, $docid);
     return $docid;
 }
 
+# no parameter shall be missing, just pass undef
 sub RegNewFile {
-    my ($self, $ext, $docid, $desc, $is_add, $hash, $daddrs, $ver, $uname) = @_;
-    return $self->RegNewFileTime($ext, $docid, $desc, $is_add, -1, $hash, $daddrs, $ver, $uname);
+    my ($self, $ext, $docid, $desc, $is_add, $fsrc, $hash, $daddrs, 
+        $ver, $uname, $perr) = @_;
+    return $self->RegNewFileTime($ext, $docid, $desc, $is_add, -1, 
+                                 $fsrc, $hash, $daddrs, $ver, $uname, $perr);
 }
 
+# no parameter shall be missing, just pass undef
 sub RegNewFileTime {
-    my ($self, $ext, $docid, $desc, $is_add, $uptime, $hash, $daddrs, $ver, $uname) = @_;
+    my ($self, $ext, $docid, $desc, $is_add, $uptime, $fsrc, $hash, $daddrs, 
+        $ver, $uname, $perr) = @_;
     if (! defined($is_add)) {$is_add = TRUE; } # Adding mode
-    if (! $self->_check_version_value($ver)) {return undef; }
+    if (! $self->_check_version_value($ver)) {
+        if (defined($perr)) {$$perr = 'version_not_numeric'; }
+        return undef;
+    }
     my $fileid = undef;
     if (! defined($uname)) {$uname = PSMT->user()->get_uid(); }
     my $srcip = PSMT::Util->IpAddr();
     my $hashcnt = 0;
     my $dbh = PSMT->dbh;
-    $dbh->db_lock_tables('docinfo WRITE');
+    $dbh->db_transaction_start();
+    # could be double, but check
+    if (defined($self->CheckFileHash($hash))) {
+        $dbh->db_transaction_rollback();
+        if (defined($perr)) {$$perr = 'file_hash_match'; }
+        return undef;
+    }
     my $sth;
+    $sth = $dbh->prepare('SELECT * FROM docinfo WHERE fileid = ?');
     while (! defined($fileid)) {
         $fileid = $self->GetHashString($docid . $uname . $srcip . $desc . $hashcnt);
-        $sth = $dbh->prepare('SELECT * FROM docinfo WHERE fileid = ?');
         $sth->execute($fileid);
         if ($sth->rows != 0) {
             $fileid = undef;
             $hashcnt += 1;
-            if ($hashcnt > 20) {return undef; }
+            if ($hashcnt > 20) {
```

Comment:
better to be configured value (in PSMT::Constant?), but could be another ticket.